bindings/azure/openai/openai.go

```diff
@@ -0,0 +1,332 @@
+		return nil, fmt.Errorf("invalid request: metadata is required")
+	}
+
+	startTime := time.Now().UTC()
+	resp = &bindings.InvokeResponse{
+		Metadata: map[string]string{
+			"operation":  string(req.Operation),
+			"start-time": startTime.Format(time.RFC3339Nano),
+		},
+	}
+
+	switch req.Operation { //nolint:exhaustive
+	case CompletionOperation:
+		response, err := p.completion(ctx, req.Data, req.Metadata)
+		if err != nil {
+			return nil, fmt.Errorf("error performing completion: %w", err)
+		}
+		responseAsBytes, _ := json.Marshal(response)
+		resp.Data = responseAsBytes
+
+	case ChatCompletionOperation:
+		response, err := p.chatCompletion(ctx, req.Data, req.Metadata)
+		if err != nil {
+			return nil, fmt.Errorf("error performing chat completion: %w", err)
+		}
+		responseAsBytes, _ := json.Marshal(response)
+		resp.Data = responseAsBytes
+
+	default:
+		return nil, fmt.Errorf(
+			"invalid operation type: %s. Expected %s, %s",
+			req.Operation, CompletionOperation, ChatCompletionOperation,
+		)
+	}
+
+	endTime := time.Now().UTC()
+	resp.Metadata["end-time"] = endTime.Format(time.RFC3339Nano)
+	resp.Metadata["duration"] = endTime.Sub(startTime).String()
+
+	return resp, nil
+}
+
+func (s *ChatSettings) Decode(in any) error {
+	return config.Decode(in, s)
+}
+
+func (p *AzOpenAI) completion(ctx context.Context, message []byte, metadata map[string]string) (response []*azopenai.Choice, err error) {
+	var prompt Prompt
+	err = json.Unmarshal(message, &prompt)
+	if err != nil {
+		return []*azopenai.Choice{}, fmt.Errorf("error unmarshalling the input object: %w", err)
+	}
+
+	messageSettings := make(map[string]any)
+	err = json.Unmarshal(message, &messageSettings)
+	if err != nil {
+		return []*azopenai.Choice{}, fmt.Errorf("error unmarshalling settings from the input object: %w", err)
+	}
+
+	if prompt.Prompt == "" {
+		return []*azopenai.Choice{}, fmt.Errorf("prompt is required for completion operation")
+	}
+
+	// Refer to the following link for more information on the parameters and their default values:
+	// https://platform.openai.com/docs/api-reference/completions/create
+	settings := ChatSettings{
+		Temperature:      1.0,
+		TopP:             1.0,
+		MaxTokens:        16,
+		N:                1,
+		PresencePenalty:  0.0,
+		FrequencyPenalty: 0.0,
+	}
+
+	err = settings.Decode(messageSettings)
+	if err != nil {
+		return []*azopenai.Choice{}, fmt.Errorf("error decoding the parameters: %w", err)
+	}
+
+	resp, err := p.client.GetCompletions(ctx, azopenai.CompletionsOptions{
+		Prompt:      []*string{&prompt.Prompt},
+		MaxTokens:   &settings.MaxTokens,
+		Temperature: &settings.Temperature,
+		TopP:        &settings.TopP,
+		N:           &settings.N,
+	}, nil)
+	if err != nil {
+		return []*azopenai.Choice{}, fmt.Errorf("error getting completion api: %w", err)
+	}
+
+	if len(resp.Completions.Choices) == 0 {
+		return []*azopenai.Choice{}, fmt.Errorf("error getting response from completion api: %w", err)
```

Comment:
Should this be an error? Or should we just return an empty response?

Also, you are using `fmt.Errorf` with `err` but here `err == nil`

Reply:
> Should this be an error? Or should we just return an empty response?

Let's return an empty response because anyway if the API threw an error then it should be already caught in the line above this. In any case the length of the response should not be 0.